### PR TITLE
feat: adjust witness size limits

### PR DESCRIPTION
**Pull request by pugachAG**
This PR includes the following:
* Set size limit for compressed witness to 32MiB. This is at least twice as big as the witness we expected to handle, so it give us quite some safety margin.
* Set size limit for the compressed witness to 64MiB. Same as above, this values give us pretty big margin on top of witness sizes we expect.
* Reduce partial witness cache size to 40 entries to limit memory usage at ~1.3GiB. 40 entries should be more than enough to handle all kinds of crazy chain forks 🙃

### CODE REVIEW

**Review by shreyan-gupta — APPROVED** [review]
Verifying, we have both `MAX_COMPRESSED_STATE_WITNESS_SIZE` and `MAX_RAW_STATE_WITNESS_SIZE` to take into account decoding size bomb thingi issue?

**Comment by pugachAG** [comment]
@shreyan-gupta only `MAX_RAW_STATE_WITNESS_SIZE` is used for that (in fact that is the only place where that is currently used). `MAX_COMPRESSED_STATE_WITNESS_SIZE` limit is applied before that when verifying partial witness. 
So yeah, effectively both limits are currently applied.

File: core/primitives/src/stateless_validation.rs [hunk 1]
```diff
@@ -16,8 +16,13 @@ use near_primitives_core::hash::CryptoHash;
...
 use near_primitives_core::version::PROTOCOL_VERSION;
 
-// The value here is the same as NETWORK_MESSAGE_MAX_SIZE_BYTES.
-pub const MAX_CHUNK_STATE_WITNESS_SIZE: ByteSize = ByteSize::mib(512);
+/// Represents max allowed size of the compressed state witness,
+/// corresponds to EncodedChunkStateWitness struct size.
+pub const MAX_COMPRESSED_STATE_WITNESS_SIZE: ByteSize = ByteSize::mib(32);
+
+/// Represents max allowed size of the raw (not compressed) state witness,
+/// corresponds to the size of borsh-serialized ChunkStateWitness.
```
Comment (tayfunelmas) [inline core/primitives/src/stateless_validation.rs:24]: nit. I think MAX_UNCOMPRESSED_STATE_WITNESS_SIZE could be clearer. or maybe MAX_STATE_WITNESS_SIZE? "raw" is ambiguous.

**Review by tayfunelmas — APPROVED** [review]
Can we even reduce it further like 16M and 32M respectively? maybe experiment with it later, do not want to block this one as it also clarifies the constant names etc.

File: core/primitives/src/stateless_validation.rs (hunk 1 above)
Comment (pugachAG) [reply core/primitives/src/stateless_validation.rs:24]: agree, done in f91fc2cd6

**Review by jancionear — APPROVED** [review]
LGTM, we with all the limits we don't expect the witness to be larger than 16MB, so 32MB should give us a safe margin.

**Comment by jancionear** [comment]
> Can we even reduce it further like 16M and 32M respectively? maybe experiment with it later, do not want to block this one as it also clarifies the constant names etc.

I would leave it at 32/64, it's safer this way. There might be some weird case where the witness gets above 16MB and with 16MB limit the network would stall. 32MB is safer, and doesn't really hurt.

**Comment by pugachAG** [comment]
@tayfunelmas Yes, I think we could reduce it further. For that we would need to carefully consider all current size limits in place (storage proof, transactions, current WIP with receipts). Currently selected values still have a lot of margin to be on the safe side, but I want to merge it ASAP to replace the current limit of 512MiB with something more reasonable.

**Pull request merged**